File: app/src/main/java/com/gproust/sprout/ui/onboarding/OnboardingScreen.kt
package com.gproust.sprout.ui.onboarding

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.gproust.sprout.R
import com.gproust.sprout.data.local.DeliveryType
import com.gproust.sprout.ui.common.ChoiceChips
import com.gproust.sprout.ui.common.DatePickerField
import com.gproust.sprout.ui.common.FieldLabel
import com.gproust.sprout.ui.common.label

@Composable
fun OnboardingScreen(
    onFinish: (
        name: String,
        gaveBirth: Boolean,
        breastfeeding: Boolean,
        deliveryType: DeliveryType?,
        babyName: String,
        birthDate: Long,
    ) -> Unit,
) {
    var step by remember { mutableIntStateOf(0) }
    var name by remember { mutableStateOf("") }
    var babyName by remember { mutableStateOf("") }
    var birthDate by remember { mutableLongStateOf(System.currentTimeMillis()) }
    var gaveBirth by remember { mutableStateOf(false) }
    var breastfeeding by remember { mutableStateOf(false) }
    var deliveryType by remember { mutableStateOf<DeliveryType?>(null) }

    Box(Modifier.fillMaxSize().padding(24.dp)) {
        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState()),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Spacer(Modifier.height(32.dp))
            when (step) {
                0 -> WelcomeStep(onNext = { step = 1 })
                1 -> AboutYouStep(
                    name = name,
                    onName = { name = it },
                    onBack = { step = 0 },
                    onNext = { step = 2 },
                )
                2 -> BabyStep(
                    babyName = babyName,
                    onBabyName = { babyName = it },
                    birthDate = birthDate,
                    onBirthDate = { birthDate = it },
                    onBack = { step = 1 },
                    onNext = { step = 3 },
                )
                else -> CareStep(
                    babyName = babyName,
                    gaveBirth = gaveBirth,
                    onGaveBirth = { gaveBirth = it },
                    deliveryType = deliveryType,
                    onDeliveryType = { deliveryType = it },
                    breastfeeding = breastfeeding,
                    onBreastfeeding = { breastfeeding = it },
                    onBack = { step = 2 },
                    onFinish = {
                        onFinish(
                            name,
                            gaveBirth,
                            breastfeeding,
                            if (gaveBirth) deliveryType else null,
                            babyName,
                            birthDate,
                        )
                    },
                )
            }
        }
    }
}

@Composable
private fun WelcomeStep(onNext: () -> Unit) {
    Text("🌱", style = MaterialTheme.typography.displayLarge)
    Spacer(Modifier.height(16.dp))
    Text(
        stringResource(R.string.onboarding_welcome_title),
        style = MaterialTheme.typography.headlineMedium,
        fontWeight = FontWeight.Bold,
    )
    Spacer(Modifier.height(12.dp))
    Text(
        stringResource(R.string.onboarding_welcome_body),
        style = MaterialTheme.typography.bodyLarge,
        textAlign = TextAlign.Center,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(Modifier.height(32.dp))
    Button(onClick = onNext, modifier = Modifier.fillMaxWidth()) {
        Text(stringResource(R.string.onboarding_get_started))
    }
}

@Composable
private fun AboutYouStep(
    name: String,
    onName: (String) -> Unit,
    onBack: () -> Unit,
    onNext: () -> Unit,
) {
    Text(
        stringResource(R.string.onboarding_about_title),
        style = MaterialTheme.typography.headlineSmall,
        fontWeight = FontWeight.Bold,
    )
    Spacer(Modifier.height(8.dp))
    Text(
        stringResource(R.string.onboarding_about_subtitle),
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center,
    )
    Spacer(Modifier.height(24.dp))
    OutlinedTextField(
        value = name,
        onValueChange = onName,
        label = { Text(stringResource(R.string.onboarding_your_name)) },
        singleLine = true,
        modifier = Modifier.fillMaxWidth(),
    )
    Spacer(Modifier.height(32.dp))
    StepButtons(
        onBack = onBack,
        nextLabel = stringResource(R.string.action_next),
        nextEnabled = name.isNotBlank(),
        onNext = onNext,
    )
}

@Composable
private fun BabyStep(
    babyName: String,
    onBabyName: (String) -> Unit,
    birthDate: Long,
    onBirthDate: (Long) -> Unit,
    onBack: () -> Unit,
    onNext: () -> Unit,
) {
    Text(
        stringResource(R.string.onboarding_baby_title),
        style = MaterialTheme.typography.headlineSmall,
        fontWeight = FontWeight.Bold,
    )
    Spacer(Modifier.height(8.dp))
    Text(
        stringResource(R.string.onboarding_baby_subtitle),
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center,
    )
    Spacer(Modifier.height(24.dp))
    OutlinedTextField(
        value = babyName,
        onValueChange = onBabyName,
        label = { Text(stringResource(R.string.field_baby_name)) },
        singleLine = true,
        modifier = Modifier.fillMaxWidth(),
    )
    FieldLabel(stringResource(R.string.field_date_of_birth))
    DatePickerField(label = stringResource(R.string.picker_born), millis = birthDate, onChange = onBirthDate)
    Spacer(Modifier.height(32.dp))
    StepButtons(
        onBack = onBack,
        nextLabel = stringResource(R.string.action_next),
        nextEnabled = true,
        onNext = onNext,
    )
}

@Composable
private fun CareStep(
    babyName: String,
    gaveBirth: Boolean,
    onGaveBirth: (Boolean) -> Unit,
    deliveryType: DeliveryType?,
    onDeliveryType: (DeliveryType?) -> Unit,
    breastfeeding: Boolean,
    onBreastfeeding: (Boolean) -> Unit,
    onBack: () -> Unit,
    onFinish: () -> Unit,
) {
    val context = LocalContext.current
    val defaultBaby = stringResource(R.string.onboarding_default_baby)
    val who = babyName.trim().ifBlank { defaultBaby }
    Text(
        stringResource(R.string.onboarding_care_title, who),
        style = MaterialTheme.typography.headlineSmall,
        fontWeight = FontWeight.Bold,
    )
    Spacer(Modifier.height(8.dp))
    Text(
        stringResource(R.string.onboarding_care_subtitle),
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center,
    )
    Spacer(Modifier.height(24.dp))
    ToggleRow(
        label = stringResource(R.string.onboarding_gave_birth_q, who),
        help = stringResource(R.string.onboarding_gave_birth_help),
        checked = gaveBirth,
        onChecked = onGaveBirth,
    )
    if (gaveBirth) {
        FieldLabel(stringResource(R.string.onboarding_birth_how))
        ChoiceChips(
            options = DeliveryType.entries,
            selected = deliveryType,
            onSelect = { onDeliveryType(if (deliveryType == it) null else it) },
            labelOf = { it.label(context) },
        )
        Spacer(Modifier.height(8.dp))
    }
    ToggleRow(
        label = stringResource(R.string.onboarding_breastfeeding_q, who),
        help = stringResource(R.string.onboarding_breastfeeding_help),
        checked = breastfeeding,
        onChecked = onBreastfeeding,
    )
    Spacer(Modifier.height(32.dp))
    StepButtons(
        onBack = onBack,
        nextLabel = stringResource(R.string.onboarding_all_done),
        nextEnabled = true,
        onNext = onFinish,
    )
}

@Composable
private fun ToggleRow(
    label: String,
    help: String,
    checked: Boolean,
    onChecked: (Boolean) -> Unit,
) {
    Row(
        Modifier.fillMaxWidth().padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(label, style = MaterialTheme.typography.bodyLarge)
            Text(
                help,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Switch(checked = checked, onCheckedChange = onChecked)
    }
}

@Composable
private fun StepButtons(
    onBack: () -> Unit,
    nextLabel: String,
    nextEnabled: Boolean,
    onNext: () -> Unit,
) {
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        OutlinedButton(onClick = onBack, modifier = Modifier.weight(1f)) {
            Text(stringResource(R.string.action_back))
        }
        Button(onClick = onNext, enabled = nextEnabled, modifier = Modifier.weight(1f)) {
            Text(nextLabel)
        }
    }
}

File: ios/Sprout/Sources/Screens/Onboarding/OnboardingScreen.swift
import SproutData
import SwiftUI

/// The first five minutes. From `ui/onboarding/OnboardingScreen.kt`.
///
/// Four steps, and each one earns its place: the welcome says what Sprout is and
/// what it does not do, the parent's own answers decide which questions the
/// wellbeing entries will offer (BDR-0001), the baby is what everything else
/// hangs off, and the care questions are asked *after* the baby has a name so
/// they can use it.
///
/// Nothing is written until the last step. A parent who backs out at step three
/// has left nothing behind, which is what makes "you can change all of this
/// later" true rather than reassuring.
struct OnboardingScreen: View {

    /// Everything the flow collected, handed over in one piece.
    struct Answers {
        var name = ""
        var babyName = ""
        var birthDate = Clock.millis
        var gaveBirth = false
        var breastfeeding = false
        var deliveryType: DeliveryType? = nil
        var trackWellbeing = true
    }

    let onFinish: (Answers) -> Void

    @State private var step = 0
    @State private var answers = Answers()

    var body: some View {
        ScrollView {
            VStack(spacing: Spacing.regular) {
                Spacer().frame(height: Spacing.section)
                switch step {
                case 0: welcome
                case 1: aboutYou
                case 2: baby
                default: care
                }
            }
            .frame(maxWidth: .infinity)
            .padding(Spacing.loose)
        }
        .sproutStyle()
        .accessibilityIdentifier("onboarding")
    }

    // MARK: - The steps

    private var welcome: some View {
        VStack(spacing: Spacing.snug) {
            Text("🌱").font(.system(size: 64))
            Text(Str.t("onboarding_welcome_title"))
                .font(.title.weight(.bold))
                .multilineTextAlignment(.center)
            Text(Str.t("onboarding_welcome_body"))
                .font(.body)
                .foregroundStyle(SproutColor.onSurfaceVariant)
                .multilineTextAlignment(.center)
            Button(Str.t("onboarding_get_started")) { step = 1 }
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)
                .padding(.top, Spacing.loose)
                .accessibilityIdentifier("onboarding-start")
        }
    }

    private var aboutYou: some View {
        VStack(alignment: .leading, spacing: Spacing.snug) {
            heading(Str.t("onboarding_about_title"), Str.t("onboarding_about_subtitle"))

            TextField(Str.t("onboarding_your_name"), text: $answers.name)
                .textFieldStyle(.roundedBorder)
                .textInputAutocapitalization(.words)

            // Asked here, with the name, because both are about the parent
            // rather than the baby — and because a parent who doesn't want the
            // check-in should be able to say so before Sprout starts offering
            // it, not after.
            ToggleRow(
                label: Str.t("onboarding_checkin_q"),
                help: Str.t("onboarding_checkin_help"),
                isOn: $answers.trackWellbeing
            )

            stepButtons(
                back: { step = 0 },
                nextLabel: Str.t("action_next"),
                // The one required answer in the whole flow: the check-in and
                // the report both address the parent by name.
                nextEnabled: !answers.name.trimmingCharacters(in: .whitespaces).isEmpty,
                next: { step = 2 }
            )
        }
    }

    private var baby: some View {
        VStack(alignment: .leading, spacing: Spacing.snug) {
            heading(Str.t("onboarding_baby_title"), Str.t("onboarding_baby_subtitle"))

            TextField(Str.t("field_baby_name"), text: $answers.babyName)
                .textFieldStyle(.roundedBorder)
                .textInputAutocapitalization(.words)

            FieldLabel(Str.t("field_date_of_birth"))
            DateField(label: Str.t("picker_born"), millis: $answers.birthDate)

            stepButtons(
                back: { step = 1 },
                nextLabel: Str.t("action_next"),
                // Not required: a baby who has not arrived yet, or has not been
                // named yet, is a real case. The profile is created without one
                // and the dashboard asks again.
                nextEnabled: true,
                next: { step = 3 }
            )
        }
    }

    private var care: some View {
        let named = answers.babyName.trimmingCharacters(in: .whitespaces)
        // "your baby" when they have not been named yet, so the questions below
        // read as sentences either way.
        let who = named.isEmpty ? Str.t("onboarding_default_baby") : named

        return VStack(alignment: .leading, spacing: Spacing.snug) {
            // Without the check-in these answers still have a job: they decide
            // which body questions the wellbeing entries offer. Say that instead
            // of promising a daily check-in the parent just turned down.
            heading(
                Str.t("onboarding_care_title", who),
                answers.trackWellbeing
                    ? Str.t("onboarding_care_subtitle")
                    : Str.t("onboarding_care_subtitle_no_checkin")
            )

            ToggleRow(
                label: Str.t("onboarding_gave_birth_q", who),
                help: Str.t("onboarding_gave_birth_help"),
                isOn: $answers.gaveBirth
            )

            if answers.gaveBirth {
                FieldLabel(Str.t("onboarding_birth_how"))
                ChoiceChips(
                    options: DeliveryType.allCases,
                    selection: $answers.deliveryType,
                    label: \.label,
                    allowsDeselection: true
                )
            }

            ToggleRow(
                label: Str.t("onboarding_breastfeeding_q", who),
                help: Str.t("onboarding_breastfeeding_help"),
                isOn: $answers.breastfeeding
            )

            stepButtons(
                back: { step = 2 },
                nextLabel: Str.t("onboarding_all_done"),
                nextEnabled: true,
                next: {
                    var finished = answers
                    // The delivery question was only asked of a parent who gave
                    // birth; keeping an answer they never gave would put it in
                    // the wellbeing questions and the report.
                    if !finished.gaveBirth { finished.deliveryType = nil }
                    onFinish(finished)
                },
                nextIdentifier: "onboarding-finish"
            )
        }
    }

    // MARK: - Pieces

    @ViewBuilder
    private func heading(_ title: String, _ subtitle: String) -> some View {
        VStack(spacing: Spacing.tight) {
            Text(title)
                .font(.title2.weight(.bold))
                .multilineTextAlignment(.center)
            Text(subtitle)
                .font(.callout)
                .foregroundStyle(SproutColor.onSurfaceVariant)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(.bottom, Spacing.tight)
    }

    @ViewBuilder
    private func stepButtons(
        back: @escaping () -> Void,
        nextLabel: String,
        nextEnabled: Bool,
        next: @escaping () -> Void,
        nextIdentifier: String? = nil
    ) -> some View {
        HStack(spacing: Spacing.snug) {
            Button(Str.t("action_back"), action: back)
                .buttonStyle(.bordered)
                .frame(maxWidth: .infinity)
            Button(nextLabel, action: next)
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)
                .disabled(!nextEnabled)
                .accessibilityIdentifier(nextIdentifier ?? "onboarding-next")
        }
        .padding(.top, Spacing.loose)
    }
}

/// A switch with the question above it and the reason underneath.
///
/// The help line is not decoration: every one of these decides what Sprout will
/// ask later, and a parent answering at 3 a.m. deserves to know which.
private struct ToggleRow: View {
    let label: String
    let help: String
    @Binding var isOn: Bool

    var body: some View {
        Toggle(isOn: $isOn) {
            VStack(alignment: .leading, spacing: 2) {
                Text(label).font(.body)
                Text(help)
                    .font(.caption)
                    .foregroundStyle(SproutColor.onSurfaceVariant)
            }
        }
        .padding(.vertical, Spacing.hairline)
    }
}

/// Writes what onboarding collected. From `StartupViewModel.completeOnboarding`.
///
/// **The baby first, then the profile.** `addBaby` makes the first baby active
/// only when a profile already exists, and at this point one does not — so the
/// selection is set here, explicitly, by saving the profile with the new baby's
/// id. Doing it the other way round leaves a phone with a baby and nothing
/// selected, which opens on an empty dashboard.
///
/// A blank baby name writes no baby at all rather than one called "". The
/// dashboard then asks again, which is the right prompt for a parent who set
/// Sprout up before the birth.
func completeOnboarding(
    _ answers: OnboardingScreen.Answers,
    into repository: SproutRepository
) throws {
    let babyName = answers.babyName.trimmingCharacters(in: .whitespaces)
    let babyId = babyName.isEmpty
        ? nil
        : try repository.addBaby(name: babyName, birthDate: answers.birthDate)

    try repository.saveParentProfile(
        ParentProfile(
            name: answers.name.trimmingCharacters(in: .whitespaces),
            gaveBirth: answers.gaveBirth,
            breastfeeding: answers.breastfeeding,
            deliveryType: answers.deliveryType,
            trackWellbeing: answers.trackWellbeing,
            lastCheckIn: nil,
            activeBabyId: babyId
        )
    )
}
